---
title: "Regional weaning weight GWAS annotations: High Plains vs. Southeast vs. Fescue Belt"
author: "Harly Durbin"
output:
  html_document:
    toc: true
    toc_depth: 3
    df_print: paged
    code_folding: hide 
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE)
library(readr)
library(dplyr)
library(tidyr)
library(stringr)
library(rlang)
library(glue)
library(ggplot2)
library(lubridate)
library(forcats)
library(magrittr)
library(purrr)
library(qvalue)
library(tidylog)
library(GALLO)
library(gprofiler2)


source(here::here("source_functions/region_key.R"))
source(here::here("source_functions/read_snp_sol.R"))
source(here::here("source_functions/regions_manhattan.R"))

options(scipen=999)
```

# Notes & questions

# Setup 

* SNP solutions & p-values from 5 GWAS:
    + Southeast, spring & fall calving
    + Fescue Belt, spring & fall calving
    + High Plains, spring & fall calving

```{r}

snp_sol238 <-
  purrr::map2(.x = c(2, 8, 3),
              .y = c("2_all", "8_all", "3_all"),
              ~ read_snp_sol_ww(r1 = .x, analysis = .y)) %>%
  reduce(bind_rows)

```

```{r}
pass <-
  snp_sol238 %>% 
  filter(neglog10p > 5) %>% 
  select(region, effect, CHR = chr, BP = pos)
  
```

```{r}
pass %>% 
  group_by(region, effect) %>% 
  tally(sort = TRUE)
```


## GTF & GFF files

```{r}
ensembl_genes <-
  import_gff_gtf(db_file = here::here("data/raw_data/Bos_taurus.UMD3.1.94.chr.gtf"),
                 file_type = "gtf")
```

```{r}
qtl <-
  import_gff_gtf(db_file = here::here("data/raw_data/qdwnldLTLSFVKSNW.txt"),
                 file_type = "gff")
```

# Annotations

## Genes

```{r}
dir_gene50k <-
  map_dfr(.x = c(2, 3, 8),
          ~ pass %>% 
            filter(effect == "direct") %>% 
            filter(region == .x) %>% 
            select(-region, -effect) %>% 
            find_genes_qtls_around_markers(db_file = ensembl_genes,
                                           marker_file = .,
                                           method = c("gene"),
                                           marker = c("snp"),
                                           interval = 50000) %>% 
            mutate(region = .x))
```

```{r}
mat_gene50k <-
  map_dfr(.x = c(2, 3, 8),
          ~ pass %>% 
            filter(effect == "maternal") %>% 
            filter(region == .x) %>% 
            select(-region, -effect) %>% 
            find_genes_qtls_around_markers(db_file = ensembl_genes,
                                           marker_file = .,
                                           method = c("gene"),
                                           marker = c("snp"),
                                           interval = 50000) %>% 
            mutate(region = .x))
```

## QTL

```{r}
dir_qtl50k <-
  map_dfr(.x = c(2, 3, 8),
          ~ pass %>% 
            filter(effect == "direct") %>% 
            filter(region == .x) %>% 
            select(-region, -effect) %>% 
            find_genes_qtls_around_markers(db_file = qtl,
                                           marker_file = .,
                                           method = c("qtl"),
                                           marker = c("snp"),
                                           interval = 50000) %>% 
            mutate(region = .x))
```

```{r}
mat_qtl50k <-
  map_dfr(.x = c(2, 3, 8),
          ~ pass %>% 
            filter(effect == "maternal") %>% 
            filter(region == .x) %>% 
            select(-region, -effect) %>% 
            find_genes_qtls_around_markers(db_file = qtl,
                                           marker_file = .,
                                           method = c("qtl"),
                                           marker = c("snp"),
                                           interval = 50000) %>% 
            mutate(region = .x))
```

# Enrichment

## Genes

```{r}

dir_gene_enrich50k <-
  purrr::map(.x = c(2, 3, 8),
             ~ dir_gene50k %>% 
               filter(region == .x) %>% 
               mutate(id = if_else(!is.na(gene_name), gene_name, gene_id)) %>%
               distinct(id) %>% 
               pull(id) %>% 
               as.list() %>% 
               purrr::set_names() %>% 
               gost(organism = "btaurus",
                    significant = TRUE))
  
```

```{r}

mat_gene_enrich50k <-
  purrr::map(.x = c(2, 3, 8),
             ~ mat_gene50k %>% 
               filter(region == .x) %>% 
               mutate(id = if_else(!is.na(gene_name), gene_name, gene_id)) %>%
               distinct(id) %>% 
               pull(id) %>% 
               as.list() %>% 
               purrr::set_names() %>% 
               gost(organism = "btaurus",
                    significant = TRUE))
  
```

```{r}
mat_go <-
  purrr::map2_dfr(.x = c(1, 2, 3),
                .y = c(2, 3, 8),
                ~ mat_gene_enrich50k[[.x]]$result %>% 
                  mutate(region = .y)) %>% 
  select(region, term_name, p_value, everything())
```

## QTL

```{r}

dir_qtl_enrich50k <-
  purrr::map_dfr(.x = c(2, 3, 8),
                 ~ dir_qtl50k %>% 
                   filter(region == .x) %>% 
                   select(-region) %>% 
                   qtl_enrich(qtl_db = qtl,
                              qtl_file = .,
                              qtl_type = "Name",
                              enrich_type = "genome",
                              padj = "BH") %>% 
                   mutate(region = .x))
  
```

```{r}

mat_qtl_enrich50k <-
  purrr::map_dfr(.x = c(2, 3, 8),
                 ~ mat_qtl50k %>% 
                   filter(region == .x) %>% 
                   select(-region) %>% 
                   qtl_enrich(qtl_db = qtl,
                              qtl_file = .,
                              qtl_type = "Name",
                              enrich_type = "genome",
                              padj = "BH") %>% 
                   mutate(region = .x))
  
```

# Genes near SNPs

## Direct

```{r}
dir_gene50k %>% 
  filter(CHR == 7)
```

```{r}
dir_gene50k %>% 
  filter(CHR == 20) %>% 
  group_by(gene_name) %>% 
  tally(sort = TRUE)
```


### Fescue Belt

* Chromosome 6: known growth QTL
    + **SLIT2**, **KCNIP4**
* Chromosome 3: known growth QTL
    + OR6N1
    + OR6K6
    + OR6K3
    + OR6K2
        - Implicated in wisent adaptation under selection olfactory, 1:1 ortholog with cattle
    + SPTA1

```{r}
dir_gene50k %>% 
  filter(CHR %in% c(3, 6)) %>% 
  select(gene_name, region, everything())
```

### Southeast

* Chromosome 14: known growth QTL
    + **FBXO32**, WDYHV1, **ATAD2**, 
* Chromosome 15: 
    + METTL15

```{r}
dir_gene50k %>% 
  filter(CHR %in% c(14, 15)) %>% 
  select(gene_name, region,everything())
```

## Region 2 maternal

```{r}
mat_gene50k %>% 
  filter(region == 2) %>% 
  left_join(snp_sol238 %>% 
              filter(region == 2 & effect == "maternal") %>% 
              select(CHR = chr, BP = pos, neglog10p)) %>% 
  arrange(desc(neglog10p)) %>% 
  select(neglog10p, gene_name, everything())

```

```{r}
mat_gene50k %>% 
  filter(region == 2) %>% 
  group_by(chr, gene_id, gene_name, gene_biotype) %>% 
  tally(sort = TRUE)
```

```{r}
mat_gene50k %>% 
  filter(region == 2) %>% 
  filter(chr == 23) %>% 
  left_join(snp_sol238 %>% 
              filter(region == 2 & effect == "maternal" & chr == 23) %>% 
              select(CHR = chr, BP = pos, neglog10p)) %>% 
  arrange(desc(neglog10p)) %>% 
  select(neglog10p, everything())

```

## Region 8 maternal

```{r}
mat_gene50k %>% 
  filter(region == 8) %>% 
  filter(chr == 1) %>% 
  left_join(snp_sol238 %>% 
              filter(region == 8 & effect == "maternal" & chr == 1) %>% 
              select(CHR = chr, BP = pos, neglog10p)) %>% 
  arrange(desc(neglog10p)) %>% 
  select(neglog10p, everything())

```

```{r}
mat_gene50k %>% 
  filter(region == 8) %>% 
  group_by(chr, gene_id, gene_name, gene_biotype) %>% 
  tally(sort = TRUE)
```


# QTL enrichment results

## Direct effect

```{r}
dir_qtl_enrich50k %>% 
  filter(0.05 > adj.pval) %>% 
  filter(N_QTLs > 1) %>% 
  left_join(dir_qtl50k %>% 
              select(QTL = trait_ID, CHR, BP, region)) %>% 
  left_join(dir_gene50k %>% 
              select(gene_name, CHR, BP, region)) %>% 
  filter(!is.na(gene_name)) %>% 
  distinct(QTL, adj.pval, gene_name, region) %>% 
  arrange(region, adj.pval, gene_name) %>% 
  group_by(QTL, adj.pval, region) %>% 
  nest(gene_name) %>% 
  ungroup() %>% 
  mutate(string = purrr::map_chr(.x = data, 
                                 ~ .x %>% 
                                   pull(gene_name) %>% 
                                   glue_collapse(sep = ", ")),
         region = case_when(region == 2 ~ "Southeast", 
                            region == 3 ~ "High Plains",
                            region == 8 ~ "Fescue Belt")) %>% 
    select(-data) %>% 
  writexl::write_xlsx("~/Desktop/dirqtl.xlsx")
  
```

```{r}
dir_qtl50k %>% 
  filter(chr == 20) %>% 
  filter(trait_ID == "Body weight (weaning)") %>% 
  left_join(dir_gene50k %>% 
              select(CHR, BP, gene_name)) %>% 
  select(gene_name, everything()) %>% 
  group_by(gene_name) %>% 
  tally(sort = TRUE)
```

### OR6 genes

```{r}
dir_qtl_enrich50k %>% 
  filter(0.05 > adj.pval) %>% 
  filter(N_QTLs > 1) %>% 
  left_join(dir_qtl50k %>% 
              select(QTL = trait_ID, CHR, BP, region)) %>% 
  left_join(dir_gene50k %>% 
              select(gene_name, CHR, BP, region)) %>% 
  filter(str_detect(gene_name, "OR6")) 
```


## Maternal effect

```{r}
mat_qtl_enrich50k %>% 
  filter(0.05 > adj.pval) %>% 
  left_join(mat_qtl50k %>% 
              select(QTL = trait_ID, CHR, BP, region)) %>% 
  left_join(mat_gene50k %>% 
              select(gene_name, CHR, BP, region)) %>% 
  filter(!is.na(gene_name)) %>% 
  distinct(QTL, adj.pval, gene_name, region) %>% 
  arrange(region, adj.pval, gene_name) %>% 
  group_by(QTL, adj.pval, region) %>% 
  nest(gene_name) %>% 
  ungroup() %>% 
  mutate(string = purrr::map_chr(.x = data, 
                                 ~ .x %>% 
                                   pull(gene_name) %>% 
                                   glue_collapse(sep = ", ")),
         region_name = case_when(region == 2 ~ "Southeast", 
                            region == 3 ~ "High Plains",
                            region == 8 ~ "Fescue Belt")) %>% 
    select(-data) %>% 
  writexl::write_xlsx("~/Desktop/matqtl.xlsx")
  
```

```{r}
mat_qtl_enrich50k %>% 
  select(dataset, adj.pval, everything()) %>% 
  filter(QTL != "") %>% 
  filter(dataset != 3) %>% 
  filter(0.05 > adj.pval) %>% 
  #filter(N_QTLs > 1) %>% 
  select(dataset, QTL_type, everything())
```


# Commentary
